src/Microsoft.ML.TorchSharp/AutoFormerV2/ObjectDetectionTrainer.cs

```diff
@@ -0,0 +1,945 @@
+                            return cursorValid;
+                    }
+                }
+
+                Updates++;
+                host.CheckAlive();
+                torch.random.manual_seed(1 + Updates);
+                torch.cuda.manual_seed(1 + Updates);
+                Model.train();
+                Model.FreezeBN();
+
+                Optimizer.zero_grad();
+
+                imageTensor = torch.cat(inputTensors);
+                targetTensor = torch.cat(targets);
+
+                var (classification, regression, anchors) = Model.forward(imageTensor);
+                var lossValue = Loss.forward(classification, regression, anchors, targetTensor);
+                lossValue.backward();
+                torch.nn.utils.clip_grad_norm_(Model.parameters(), 0.1);
+
+                Optimizer.step();
+                host.CheckAlive();
+
+                return cursorValid;
+            }
+
+            private (Tensor image, Tensor Label) PrepareData(ValueGetter<VBuffer<uint>> labelGetter, ValueGetter<MLImage> imageGetter, ValueGetter<VBuffer<float>> boundingBoxGetter)
+            {
+                using (var _ = torch.NewDisposeScope())
+                {
+                    //var image = new Image(this.imageList[idx]);
+                    MLImage image = default;
+                    imageGetter(ref image);
+                    var midTensor0 = torch.tensor(image.Pixels.ToArray());
+                    var midTensor1 = midTensor0.@float();
+                    midTensor1 = midTensor1.slice(0, 0, image.Height * image.Width * 3, 1); // TODO: FIX THIS
+                    var midTensor2 = midTensor1.reshape(1, image.Height, image.Width, 3);
+                    var midTensor3 = midTensor2.transpose(0, 3);
+                    var midTensor4 = midTensor3.reshape(3, image.Height, image.Width);
+                    var chunks = midTensor4.chunk(3, 0);
+
+                    List<Tensor> part = new List<Tensor>();
+                    part.Add(chunks[2]);
+                    part.Add(chunks[1]);
+                    part.Add(chunks[0]);
+
+                    using var midTensor = torch.cat(part, 0);
+                    using var reMidTensor = midTensor.reshape(1, 3, image.Height, image.Width);
+                    var padW = 32 - (image.Width % 32);
+                    var padH = 32 - (image.Height % 32);
+                    using var transMidTensor = torch.zeros(1, 3, image.Height + padH, image.Width + padW);
+                    transMidTensor[.., .., ..image.Height, ..image.Width] = reMidTensor / 255.0;
+                    var imageTensor = Normalize(transMidTensor);
+
+                    VBuffer<uint> labels = default;
+                    labelGetter(ref labels);
+
+                    VBuffer<float> boxes = default;
+                    boundingBoxGetter(ref boxes);
+
+                    var labelValues = labels.GetValues();
+                    var boxValues = boxes.GetValues();
+
+                    int b = 0;
```

Comment:
Might want to assert if len(labelValues) * 4 == len(boxValues) here

Reply:
Good idea. Its been added.